### PR TITLE
feat: make pending cache generic

### PR DESCRIPTION
**Pull request by MatousJobanek**
In UserSignup controller, we use a mapper together with a cache that stores unapproved (pending) UserSignups sorted by their creation date. This helps us in situations when we wouldn't have enough capacity in our cluster and the users would end up on a waitlist.
We need the same logic also for Spaces (those ones that are created directly without specifying a target cluster).

So I moved the logic to a separate package and made the functions more generic. I added also some functions for Spaces (to verify that it will work and it is generic enough)

it depends on: https://github.com/codeready-toolchain/api/pull/290

### CODE REVIEW

File: pkg/pending/cache.go [hunk 1]
```diff
@@ -0,0 +1,105 @@
+package pending
+
+import (
+	"context"
+	"sort"
+	"sync"
+
+	toolchainv1alpha1 "github.com/codeready-toolchain/api/api/v1alpha1"
+	"github.com/pkg/errors"
```
Comment (xcoulon) [inline pkg/pending/cache.go:9]: this package should be aliased as `errs` (as suggested on Slack this week)

File: pkg/pending/cache.go [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+		return nil
+	}
+	name := c.objectsByCreation[0]
+	firstExisting := c.objectType.DeepCopyObject().(client.Object)
+	if err := c.client.Get(context.TODO(), types.NamespacedName{Namespace: namespace, Name: name}, firstExisting); err != nil {
+		if apierrors.IsNotFound(err) {
+			c.objectsByCreation = c.objectsByCreation[1:]
+			return c.getFirstExisting(namespace)
+		}
+		log.Error(err, "could not get the oldest unapproved UserSignup")
```
Comment (xcoulon) [inline pkg/pending/cache.go:70]:
```suggestion
		log.Error(err, "could not get the oldest unapproved object")
```

File: pkg/pending/cache.go (hunk 1 above)
Comment (xcoulon) [inline pkg/pending/cache.go:9]: this package should be aliased as `errs` (as suggested on Slack this week)

File: pkg/pending/cache.go [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+	"github.com/pkg/errors"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+)
+
+var log = logf.Log.WithName("pending_object_cache")
+
+type GetListOfPendingObjects func(cl client.Client, labelListOption client.ListOption) ([]client.Object, error)
```
Comment (xcoulon) [inline pkg/pending/cache.go:18]:
```suggestion
type ListPendingObjects func(cl client.Client, labelListOption client.ListOption) ([]client.Object, error)
```

Comment (xcoulon) [reply pkg/pending/cache.go:18]: using a shorter name for readbility

File: pkg/pending/cache.go [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+	}
+	objects := make([]client.Object, len(userSignupList.Items))
+	for i := range userSignupList.Items {
+		userSignup := userSignupList.Items[i]
+		objects[i] = &userSignup
+	}
+	return objects, nil
+}
+
+var getListOfPendingSpaces GetListOfPendingObjects = func(cl client.Client, labelListOption client.ListOption) ([]client.Object, error) {
```
Comment (xcoulon) [inline pkg/pending/cache.go:94]: or you can just declare it as:
```suggestion
func ListPendingSpaces func(cl client.Client, labelListOption client.ListOption) ([]client.Object, error) {
```

and use this func as an argument when calling `newCache(...)`

File: pkg/pending/cache.go [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+	}
+	if firstExisting.GetLabels()[toolchainv1alpha1.StateLabelKey] != toolchainv1alpha1.StateLabelValuePending {
+		c.objectsByCreation = c.objectsByCreation[1:]
+		return c.getFirstExisting(namespace)
+	}
+
+	return firstExisting
+}
+
+var getListOfPendingUserSignups GetListOfPendingObjects = func(cl client.Client, labelListOption client.ListOption) ([]client.Object, error) {
```
Comment (xcoulon) [inline pkg/pending/cache.go:81]: or you can just declare it as:
```suggestion
func ListPendingUserSignups func(cl client.Client, labelListOption client.ListOption) ([]client.Object, error) {
```
and use this func as an argument when calling newCache(...)

File: pkg/pending/cache.go [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+var log = logf.Log.WithName("pending_object_cache")
+
+type GetListOfPendingObjects func(cl client.Client, labelListOption client.ListOption) ([]client.Object, error)
+
+type cache struct {
+	sync.RWMutex
+	objectsByCreation       []string
+	client                  client.Client
+	objectType              client.Object
+	getListOfPendingObjects GetListOfPendingObjects
```
Comment (xcoulon) [inline pkg/pending/cache.go:25]:
```suggestion
	listPendingObjects ListPendingObjects
```
(due to renaming above)

File: pkg/pending/cache.go [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+)
+
+var log = logf.Log.WithName("pending_object_cache")
+
+type GetListOfPendingObjects func(cl client.Client, labelListOption client.ListOption) ([]client.Object, error)
+
+type cache struct {
+	sync.RWMutex
+	objectsByCreation       []string
```
Comment (xcoulon) [inline pkg/pending/cache.go:22]:
```suggestion
	objectNames       []string
```

Comment (xcoulon) [reply pkg/pending/cache.go:22]: or `objectsByCreationTimestamp`, maybe? 🤔

File: pkg/pending/cache.go (hunk 2 above)
Comment (xcoulon) [reply pkg/pending/cache.go:70]: or see if the error message can contain the object type, using something like 
```
log.Error(err, fmt.Sprintf("could not get the oldest unapproved object of type '%T'", c.objectType))
``` 
for example?

File: pkg/pending/cache_test.go [hunk 8]
```diff
@@ -0,0 +1,300 @@
...
+	assert.Len(t, cache.objectsByCreation, 1)
+	assert.Equal(t, pending.Name, foundPending.GetName())
+	approve(t, cl, pending)
+
+	t.Run("won't return any since all are approved", func(t *testing.T) {
+		// when
+		foundPending := cache.getOldestPendingObject(test.HostOperatorNs)
+
+		// then
+		assert.Len(t, cache.objectsByCreation, 0)
```
Comment (xcoulon) [inline pkg/pending/cache_test.go:45]:
```suggestion
		assert.Empty(t, cache.objectsByCreation)
```

File: pkg/pending/cache.go (hunk 1 above)
Comment (MatousJobanek) [reply pkg/pending/cache.go]: done 640536aa

File: pkg/pending/cache.go (hunk 2 above)
Comment (MatousJobanek) [reply pkg/pending/cache.go]: changed 7795807d

File: pkg/pending/cache.go (hunk 1 above)
Comment (MatousJobanek) [reply pkg/pending/cache.go]: already done in 640536a

File: pkg/pending/cache.go (hunk 3 above)
Comment (MatousJobanek) [reply pkg/pending/cache.go]: renamed 994d9345

File: pkg/pending/cache.go (hunk 4 above)
Comment (MatousJobanek) [reply pkg/pending/cache.go]: renamed 994d9345

File: pkg/pending/cache.go (hunk 5 above)
Comment (MatousJobanek) [reply pkg/pending/cache.go]: renamed af36438

File: pkg/pending/cache.go (hunk 6 above)
Comment (MatousJobanek) [reply pkg/pending/cache.go]: renamed af36438

File: pkg/pending/cache.go (hunk 7 above)
Comment (MatousJobanek) [reply pkg/pending/cache.go]: this? fc356141

File: pkg/pending/cache_test.go (hunk 8 above)
Comment (MatousJobanek) [reply pkg/pending/cache_test.go]: done in a0bc941b

File: pkg/pending/cache.go (hunk 7 above)
Comment (xcoulon) [reply pkg/pending/cache.go]: yes, if you prefer that name, it works for me, too

**Comment by MatousJobanek** [comment]
@xcoulon thanks a lot for your review. I believe that I addressed all your comments

**Review by xcoulon — APPROVED** [review]

**Comment by xcoulon** [comment]
> @xcoulon thanks a lot for your review. I believe that I addressed all your comments

awesome, thanks for taking my comments into account, @MatousJobanek!

File: pkg/pending/cache_test.go [hunk 9]
```diff
@@ -0,0 +1,300 @@
...
+	waitForFinished.Wait()
+
+	// when
+	foundPending := cache.getOldestPendingObject(test.HostOperatorNs)
+
+	// then
+	assert.Nil(t, foundPending)
+}
+
+func newCache(t *testing.T, objectType client.Object, getListOfPendingObjects ListPendingObjects, initObjects ...runtime.Object) (*cache, *test.FakeClient) {
```
Comment (xcoulon) [inline pkg/pending/cache_test.go:289]:
```suggestion
func newCache(t *testing.T, objectType client.Object, listPendingObjects ListPendingObjects, initObjects ...runtime.Object) (*cache, *test.FakeClient) {
```

Comment (xcoulon) [reply pkg/pending/cache_test.go:289]: sorry, I missed that one during my first review :/

File: pkg/pending/cache_test.go [hunk 10]
```diff
@@ -0,0 +1,300 @@
...
+	s := scheme.Scheme
+	err := apis.AddToScheme(s)
+	require.NoError(t, err)
+
+	fakeClient := test.NewFakeClient(t, initObjects...)
+	return &cache{
+		client:             fakeClient,
+		objectType:         objectType,
+		listPendingObjects: getListOfPendingObjects,
```
Comment (xcoulon) [inline pkg/pending/cache_test.go:298]:
```suggestion
		listPendingObjects: listPendingObjects,
```

**Pull request merged**